Docs for app config and also modifies WordCount App to use the app config feature

gokulavasan opened the pull request:
Build : http://builds.cask.co/browse/CDAP-RBT339-1

albertshau commented on cdap-examples/WordCount/src/main/java/co/cask/cdap/examples/wordcount/WordCounter.java:
```diff
@@ -41,15 +41,21 @@
+  private final WordCount.WordCountConfig config;
+
+  public WordCounter(WordCount.WordCountConfig config) {
+    this.config = config;
```
o this works? for some reason thought we weren't supporting this yet.

albertshau replied:
its inconsistent that the other programs use the property annotation and this doesn't.

gokulavasan replied:
Oh, so this is used only in Flow. And Flow's methods are executed only during configuration time anyway. 

Property annotation is used if we want things to be available in runtime. So Flowlets, HttpServiceHandlers use them.

albertshau replied:
ok i see. Do we have the property stuff documented anywhere? If not can you open a jira? it is not very intuitive.

gokulavasan replied:
It is documented. Needs a bit more explanation, maybe? http://docs.cask.co/cdap/current/en/developers-manual/advanced/best-practices.html?highlight=property%20annotation.

albertshau commented on cdap-docs/developers-manual/source/building-blocks/applications.rst:
```diff
@@ -51,6 +51,38 @@ name.
 Names used for streams and datasets need to be unique across the CDAP instance, while
 names used for programs and services need to be unique only to the application.
 
+Application can also use a Configuration class to get a configuration during deployment time of the Application.
```
need some more detail.  For example, how is the configuration specified? Maybe link to the rest api section for more detail.

awholegunch replied:
Applications...
...to receive a configuration

awholegunch commented on cdap-docs/developers-manual/source/building-blocks/applications.rst:
```diff
@@ -51,6 +51,38 @@ name.
 names used for programs and services need to be unique only to the application.
 
+Application can also use a Configuration class to get a configuration during deployment time of the Application.
+This can be used to, for example, provide the stream, dataset to create/read from during application deployment time
```
For example, this can be used to specify|---|at application deployment time|---|a stream to be created or a dataset to be read instead of hard-coding it in... [delete "For example, "] We can modify...

albertshau commented on cdap-docs/reference-manual/source/http-restful-api/lifecycle.rst:
```diff
@@ -25,6 +25,10 @@ with the name of the JAR file as a header::
 
+configuration object as a serialized json as another header (optional) ::
+
+  X-App-Config: <Serialized JSON of the Configuration Object>
```
I think we need more detail about what this means exactly, probably with an example to illustrate

awholegunch commented on cdap-docs/reference-manual/source/http-restful-api/lifecycle.rst:
```diff
@@ -25,6 +25,10 @@ with the name of the JAR file as a header::
 
   X-Archive-Name: <JAR filename>
 
+configuration object as a serialized json as another header (optional) ::
```
JSON...(optional)::

[close up space before double colons]

awholegunch commented on cdap-examples/WordCount/src/main/java/co/cask/cdap/examples/wordcount/RetrieveCounts.java:
```diff
@@ -25,11 +25,16 @@
   @Override
   protected void configure() {
     setName(SERVICE_NAME);
     setDescription("A service to retrieve statistics, word counts and associations.");
```
statistics, word counts, and associations. [insert comma]

awholegunch commented on cdap-examples/WordCount/src/main/java/co/cask/cdap/examples/wordcount/UniqueCounter.java:
```diff
@@ -16,16 +16,37 @@
+import co.cask.cdap.api.flow.flowlet.FlowletContext;
 
 /**
  *
```
Add Javadoc?

awholegunch commented on cdap-examples/WordCount/src/main/java/co/cask/cdap/examples/wordcount/WordCount.java:
```diff
@@ -16,34 +16,86 @@
 
 /**
- * Word count sample Application.
+ * Word count sample Application. Uses a configuration class which can be used to pass in a configuration during
```
Word count sample Application. Includes a configuration class which can be used to pass in a configuration during
application deployment time. [consistent with other usages]

albertshau commented on cdap-examples/WordCount/src/main/java/co/cask/cdap/examples/wordcount/WordCount.java:
```diff
@@ -16,34 +16,86 @@
+    private String ucTable;
+    private String waTable;
+
+    public WordCountConfig() {
```
should add a comment for why this is needed

albertshau commented on cdap-examples/WordCount/src/main/java/co/cask/cdap/examples/wordcount/WordCount.java:
```diff
@@ -16,34 +16,86 @@
+      this.waTable = "wordAssocs";
+    }
+
+    public WordCountConfig(String stream, String wsTable, String wcTable, String ucTable, String waTable) {
```
similarly, comment that this is purely for unit tests and not used by CDAP

albertshau commented on cdap-examples/WordCount/src/main/java/co/cask/cdap/examples/wordcount/WordCount.java:
```diff
@@ -16,34 +16,86 @@
+   */
+  public static class WordCountConfig extends Config {
+    private String stream;
+    private String wsTable;
```
I think it's better to use the full names, can't really tell what these are just by looking at them

albertshau commented:
maybe also ask somebody who wasn't involved in working on this feature to see if it is clear to them

awholegunch commented on cdap-examples/WordCount/src/main/java/co/cask/cdap/examples/wordcount/WordCounter.java:
```diff
@@ -41,15 +41,21 @@
+    addFlowlet("associator", new WordAssociator(config.getWaTable()));
+    addFlowlet("counter", new Counter(config.getWcTable()));
+    addFlowlet("unique", new UniqueCounter(config.getUcTable()));
+    connectStream(config.getStream(), "splitter");
```
Is it confusing to have a Flowlet and a Stream with the same name? Just asking...

gokulavasan replied:
Hmm, not sure if I understand. Stream name is from config.getStream(). Flowlet name is ``splitter``.

awholegunch commented:
As you have changed ``WordCount.java``, you will need to update the hash in ``~/Source/cdap_3.1/cdap-docs/examples-manual/build.sh``...

gokulavasan commented:
@poornachandra Please take a look at the PR. Thanks!

gokulavasan commented:
@awholegunch I will update the hash before merging the PR since there might be more changes to it based on comments.

gokulavasan commented on cdap-docs/developers-manual/source/building-blocks/applications.rst:
```diff
@@ -48,9 +48,77 @@ Other components are defined using user-written classes that implement correspon
+the type parameter of ``AbstractApplication`` class. It should also extend the ``Config`` class present in CDAP API.
+The configuration is provided during application creation time through a Header in the REST API. It is available during
+configuration time through the ``getConfig()`` method in ``AbstractApplication``.
+Information about the REST call is available in Lifecycle HTTP REST API documentation.
```
@awholegunch How do I link the Lifecycle HTTP REST API documentation here?

awholegunch replied:
```
Information about the RESTful call is available in the :ref:`Lifecycle HTTP RESTful API documentation <http-restful-api-lifecycle>`.
```

awholegunch commented on cdap-docs/developers-manual/source/building-blocks/applications.rst:
```diff
@@ -48,9 +48,77 @@ Other components are defined using user-written classes that implement correspon
+        }
+      }
+
+In order to use the configuration in programs, we pass it to individual Programs using their constructor. But if
```
programs using their constructor. If

awholegunch commented on cdap-docs/developers-manual/source/building-blocks/applications.rst:
```diff
@@ -48,9 +48,77 @@ Other components are defined using user-written classes that implement correspon
+Names used for streams and datasets need to be unique across the CDAP namespace, while
 names used for programs and services need to be unique only to the application.
 
+Applications can also use a Configuration class to receive a configuration during deployment time of the Application.
```
```
Applications can use a ``Configuration`` class
```

awholegunch commented on cdap-docs/developers-manual/source/building-blocks/applications.rst:
```diff
@@ -48,9 +48,77 @@ Other components are defined using user-written classes that implement correspon
 
+Applications can also use a Configuration class to receive a configuration during deployment time of the Application.
+For example, this can be used to specify the stream, dataset to create/read from at application deployment time
+instead of hard coding it in the ``AbstractApplication``'s ``configure`` method. The configuration class needs to be
```
hard-coding

awholegunch replied:
hard-coding them

awholegunch commented on cdap-docs/developers-manual/source/building-blocks/applications.rst:
```diff
@@ -48,9 +48,77 @@ Other components are defined using user-written classes that implement correspon
 names used for programs and services need to be unique only to the application.
 
+Applications can also use a Configuration class to receive a configuration during deployment time of the Application.
+For example, this can be used to specify the stream, dataset to create/read from at application deployment time
```
For example, this can be used to specify that |---| at application deployment time |---| a stream be created or a dataset
be read from, rather than hard-coding them in the...

gokulavasan replied:
Ok not sure what is the change that I should make here. Can you paste the modified line and I will just use it.

awholegunch replied:
```
For example, this can be used to specify |---| for application deployment time |---|
a stream to be created or a dataset to be read, rather than having them hard-coded in the
 ``AbstractApplication``'s ``configure`` method.
```

awholegunch commented on cdap-docs/developers-manual/source/building-blocks/applications.rst:
```diff
@@ -48,9 +48,77 @@ Other components are defined using user-written classes that implement correspon
+Applications can also use a Configuration class to receive a configuration during deployment time of the Application.
+For example, this can be used to specify the stream, dataset to create/read from at application deployment time
+instead of hard coding it in the ``AbstractApplication``'s ``configure`` method. The configuration class needs to be
+the type parameter of ``AbstractApplication`` class. It should also extend the ``Config`` class present in CDAP API.
```
of the ``AbstractApplication`` class.

awholegunch replied:
It should also extend the ``Config`` class present in the CDAP API.

awholegunch commented on cdap-docs/developers-manual/source/building-blocks/applications.rst:
```diff
@@ -48,9 +48,77 @@ Other components are defined using user-written classes that implement correspon
+For example, this can be used to specify the stream, dataset to create/read from at application deployment time
+instead of hard coding it in the ``AbstractApplication``'s ``configure`` method. The configuration class needs to be
+the type parameter of ``AbstractApplication`` class. It should also extend the ``Config`` class present in CDAP API.
+The configuration is provided during application creation time through a Header in the REST API. It is available during
```
application deployment time through a header in the RESTful API.

awholegunch commented on cdap-docs/reference-manual/source/http-restful-api/lifecycle.rst:
```diff
@@ -32,6 +36,21 @@ and its content as the body of the request::
 Invoke the same command to update an application to a newer version.
 However, be sure to stop all of its flows, Spark and MapReduce programs before updating the application.
 
+For an application that has following configuration class::
```
For an application with a configuration class such as::

awholegunch commented on cdap-docs/reference-manual/source/http-restful-api/lifecycle.rst:
```diff
@@ -32,6 +36,21 @@ and its content as the body of the request::
+    String datasetName;
+  }
+
+we can deploy it using the following REST call::
```
we can deploy it with this RESTful call::

gokulavasan commented:
@awholegunch Addressed your comments except for this : https://github.com/caskdata/cdap/pull/3211#discussion-diff-34938688

Can you let me know what has to be done for that?

@poornachandra Please review the PR when you get a chance.

awholegunch commented on cdap-docs/developers-manual/source/building-blocks/applications.rst:
```diff
@@ -48,9 +48,78 @@ Other components are defined using user-written classes that implement correspon
+
+We can modify the ``MyApp`` class above to take in a Configuration ``MyApp.MyAppConfig``::
+
+      public class MyApp extends AbstractApplication<MyApp.MyAppConfig> {
```
This code block can be moved over so that it is indented two spaces from the edge.

poornachandra commented on cdap-docs/developers-manual/source/building-blocks/applications.rst:
```diff
@@ -48,9 +48,78 @@ Other components are defined using user-written classes that implement correspon
+Names used for streams and datasets need to be unique across the CDAP namespace, while
 names used for programs and services need to be unique only to the application.
 
+Applications can use a ``Config`` class to receive a configuration during deployment time of the Application.
```
Can you move this section after `A Typical CDAP Application` section? It feels like a lot of information to read through even before knowing what an Application is. What do you say?

Also, this section can be then named as `Deploy Time Configuration` or something like that.

awholegunch replied:
+1
Though, for a title, I read it as  " deploy 'Time Configuration' " ..."Deployment Configuration", perhaps?

...is there a difference between application deployment time and configuration time?

gokulavasan replied:
Currently (for 3.1) there is no difference between deployment time and configuration time. In 3.2, the terminology might change. We will have deployment followed by creation.

poornachandra commented on cdap-docs/developers-manual/source/building-blocks/applications.rst:
```diff
@@ -48,9 +48,78 @@ Other components are defined using user-written classes that implement correspon
+      }
+
+In order to use the configuration in programs, we pass it to individual Programs using their constructor. If
+the configuration parameter is also required during runtime, you can use the ``@Property`` annotation::
```
`If the configuration parameter is also required during runtime` is a bit confusing. It is not clear when the property passed through constructor argument is used vs property with `@Property` annotation.

poornachandra commented:
@gokulavasan Just two comments, rest of the documentation looks good.

awholegunch commented on cdap-docs/developers-manual/source/building-blocks/applications.rst:
```diff
@@ -48,9 +48,78 @@ Other components are defined using user-written classes that implement correspon
+        }
+      }
+
+In order to use the configuration in programs, we pass it to individual Programs using their constructor. If
```
we pass it to individual programs through their constructor.

awholegunch commented on cdap-docs/reference-manual/source/http-restful-api/lifecycle.rst:
```diff
@@ -32,6 +36,21 @@ and its content as the body of the request::
 Invoke the same command to update an application to a newer version.
 However, be sure to stop all of its flows, Spark and MapReduce programs before updating the application.
 
+For an application that has following configuration class such as::
```
For an application that has a configuration class such as::

awholegunch commented on cdap-docs/reference-manual/source/http-restful-api/lifecycle.rst:
```diff
@@ -25,6 +25,10 @@ with the name of the JAR file as a header::
+
+  X-App-Config: <JSON Serialization String of the Configuration Object>
+
 and its content as the body of the request::
```
The application's content is the body of the request::

awholegunch commented on cdap-docs/reference-manual/source/http-restful-api/lifecycle.rst:
```diff
@@ -25,6 +25,10 @@ with the name of the JAR file as a header::
 
   X-Archive-Name: <JAR filename>
 
+configuration object as a serialized JSON string as another header (optional)::
```
(An optional header can supply a configuration object as a serialized JSON string:)

::

[I think this will work better, as continuing the sentence as we did previous doesn't work to well with the optional component in the middle. Note the :: by itself to trigger the code formatting.]

gokulavasan commented:
@awholegunch Please take a look again. Addressed your comments.

awholegunch commented on cdap-docs/developers-manual/source/building-blocks/applications.rst:
```diff
@@ -67,3 +67,78 @@ Of course, not all components are required: it depends on the application. A min
+      this.uniqueCountTable.updateUniqueCount(word);
+    }
+  }
+
```
What about a link to the example?
```
An example demonstrating using a configuration is the :ref:`WordCount example <examples-word-count>`.
```

awholegunch commented:
I added a few files to fix the hash and references an included file. I would add that link referenced above, and then it LGTM.

gokulavasan commented:
@awholegunch Please take look again. Thanks.

awholegunch commented:
LGTM.
